src/thinblock.cpp

```diff
@@ -247,6 +248,107 @@ bool CXThinBlock::CheckBlockHeader(const CBlockHeader &block, CValidationState &
     return true;
 }
 
+bool CXThinBlock::HandleMessage(CDataStream &vRecv, CNode *pfrom, string strCommand, unsigned nHops)
+{
+    if (!pfrom->ThinBlockCapable())
+    {
+        LOCK(cs_main);
+        Misbehaving(pfrom->GetId(), 100);
+        return error("%s message received from a non thinblock node, peer=%d", strCommand, pfrom->GetId());
+    }
+
+    int nSizeThinBlock = vRecv.size();
+    CXThinBlock thinBlock;
+    vRecv >> thinBlock;
+
+    // Message consistency checking
+    if (!IsThinBlockValid(pfrom, thinBlock.vMissingTx, thinBlock.header))
+    {
+        LOCK(cs_main);
+        Misbehaving(pfrom->GetId(), 100);
+        return error("Invalid %s received", strCommand);
+    }
+
+    // Is there a previous block or header to connect with?
+    {
+        LOCK(cs_main);
+        uint256 prevHash = thinBlock.header.hashPrevBlock;
+        BlockMap::iterator mi = mapBlockIndex.find(prevHash);
+        if (mi == mapBlockIndex.end())
+        {
+            Misbehaving(pfrom->GetId(), 10);
+            return error("%s from peer %s (%d) will not connect, unknown previous block %s",
+                         strCommand, pfrom->addrName.c_str(), pfrom->id, prevHash.ToString());
+        }
+        CBlockIndex *pprev = mi->second;
+        CValidationState state;
+        if (!ContextualCheckBlockHeader(thinBlock.header, state, pprev))
+        {
+            // Thin block does not fit within our blockchain
+            Misbehaving(pfrom->GetId(), 100);
+            return error("%s from peer %s (%d) contextual error: %s", strCommand, pfrom->addrName.c_str(),
+                         pfrom->id, state.GetRejectReason().c_str());
+        }
+    }
+
+    CInv inv(MSG_BLOCK, thinBlock.header.GetHash());
+    bool fAlreadyHave = false;
+
+    if (nHops > 0)
```

Comment:
I there should be some comments regarding what hops is referring to...since we're wedding thinblocks and expedited...we should put some explanation as to why and what these hops are referring to..you could put it here or at the top of the function.